source/non_matching/fe_values.cc

```diff
@@ -0,0 +1,399 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2021 - 2021 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/base/exceptions.h>
+
+#include <deal.II/lac/block_vector.h>
+#include <deal.II/lac/la_parallel_block_vector.h>
+#include <deal.II/lac/la_parallel_vector.h>
+#include <deal.II/lac/la_vector.h>
+#include <deal.II/lac/petsc_block_vector.h>
+#include <deal.II/lac/petsc_vector.h>
+#include <deal.II/lac/trilinos_epetra_vector.h>
+#include <deal.II/lac/trilinos_parallel_block_vector.h>
+#include <deal.II/lac/trilinos_tpetra_vector.h>
+#include <deal.II/lac/trilinos_vector.h>
+#include <deal.II/lac/vector.h>
+
+#include <deal.II/non_matching/fe_values.h>
+
+#include <deal.II/numerics/fe_field_function.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace NonMatching
+{
+  RegionUpdateFlags::RegionUpdateFlags()
+    : inside(update_default)
+    , outside(update_default)
+    , surface(update_default)
+  {}
+
+
+  namespace internal
+  {
+    namespace FEValuesImplementation
+    {
+      template <int dim, class VectorType>
+      RefSpaceFEFieldFunction<dim, VectorType>::RefSpaceFEFieldFunction(
+        const DoFHandler<dim> &dof_handler,
+        const VectorType &     dof_values)
+        : dof_handler(&dof_handler)
+        , global_dof_values(&dof_values)
+      {
+        Assert(dof_handler.n_dofs() == dof_values.size(),
+               ExcDimensionMismatch(dof_handler.n_dofs(), dof_values.size()));
+      }
+
+
+
+      template <int dim, class VectorType>
+      void
+      RefSpaceFEFieldFunction<dim, VectorType>::set_active_cell(
+        const typename Triangulation<dim>::active_cell_iterator &cell)
+      {
+        Assert(
+          &cell->get_triangulation() == &dof_handler->get_triangulation(),
+          ExcMessage(
+            "The incoming cell must belong to the triangulation associated with "
+            "the DoFHandler passed to the constructor."));
+
+        const typename DoFHandler<dim>::active_cell_iterator dof_handler_cell(
+          &dof_handler->get_triangulation(),
+          cell->level(),
+          cell->index(),
+          dof_handler);
+
+        // Save the element and the local dof values, since this is what we need
+        // to evaluate the function.
+        element = &dof_handler_cell->get_fe();
```

Comment:
```suggestion
        element = &dof_handler->get_fe();
```

Reply:
I don't think this will not work, different cells might use different elements. This gives the first element in the `FECollection`.

Reply:
True!